File: packages/clients/js/src/module.ts
import * as tg from "./index.ts";
import { Resolve } from "./resolve.ts";

export class Module {
	[Resolve.atomic]: null;
	kind: Module.Kind;
	referent: tg.Referent<tg.Module.Source>;

	constructor(arg: {
		kind: Module.Kind;
		referent: tg.Referent<tg.Module.Source>;
	}) {
		this[Resolve.atomic] = null;
		this.kind = arg.kind;
		this.referent = arg.referent;
	}
}

export namespace Module {
	export type Kind =
		| "js"
		| "ts"
		| "dts"
		| "object"
		| "artifact"
		| "blob"
		| "directory"
		| "file"
		| "symlink"
		| "graph"
		| "command";

	export type Source = string | tg.Graph.Edge<tg.Object>;

	export namespace Source {
		export let toDataString = (value: tg.Module.Source): string => {
			if (typeof value === "string") {
				if (value.startsWith(".") || value.startsWith("/")) {
					return value;
				} else {
					return `./${value}`;
				}
			} else {
				return tg.Graph.Edge.toDataString(value, (object) => object.id);
			}
		};
	}

	export let toData = (value: tg.Module): tg.Module.Data => {
		return {
			kind: value.kind,
			referent: tg.Referent.toData(value.referent, (source) =>
				typeof source === "string"
					? source
					: tg.Graph.Edge.toDataString(source, (object) => object.id),
			),
		};
	};

	export let fromData = (data: tg.Module.Data): tg.Module => {
		return new tg.Module({
			kind: data.kind,
			referent: tg.Referent.fromData(data.referent, (source) => {
				if (
					typeof source === "string" &&
					(source.startsWith(".") || source.startsWith("/"))
				) {
					return source;
				} else {
					return tg.Graph.Edge.fromData(source, tg.Object.withId);
				}
			}),
		});
	};

	export let toDataString = (value: tg.Module): string => {
		let string = tg.Module.Source.toDataString(value.referent.node);
		let params = [];
		if (
			value.referent.options?.artifact !== undefined &&
			value.referent.options.artifact !== null
		) {
			params.push(
				`artifact=${encodeURIComponent(value.referent.options.artifact)}`,
			);
		}
		if (
			value.referent.options?.id !== undefined &&
			value.referent.options.id !== null
		) {
			params.push(`id=${encodeURIComponent(value.referent.options.id)}`);
		}
		if (
			value.referent.options?.location !== undefined &&
			value.referent.options.location !== null
		) {
			let location = tg.Location.toDataString(value.referent.options.location);
			params.push(`location=${encodeURIComponent(location)}`);
		}
		if (
			value.referent.options?.name !== undefined &&
			value.referent.options.name !== null
		) {
			params.push(`name=${encodeURIComponent(value.referent.options.name)}`);
		}
		if (
			value.referent.options?.path !== undefined &&
			value.referent.options.path !== null
		) {
			params.push(`path=${encodeURIComponent(value.referent.options.path)}`);
		}
		if (
			value.referent.options?.tag !== undefined &&
			value.referent.options.tag !== null
		) {
			params.push(`tag=${encodeURIComponent(value.referent.options.tag)}`);
		}
		for (let [location, token] of Object.entries(
			value.referent.options?.tokens ?? {},
		)) {
			params.push(
				`tokens[${encodeURIComponent(location)}]=${encodeURIComponent(token)}`,
			);
		}
		params.push(`kind=${encodeURIComponent(value.kind)}`);
		string += "?";
		string += params.join("&");
		return string;
	};

	export let fromDataString = (data: string): tg.Module => {
		let [nodeString, params] = data.split("?");
		tg.assert(nodeString !== undefined);
		let kind: tg.Module.Kind | undefined;
		let source: tg.Module.Source;
		if (
			typeof nodeString === "string" &&
			(nodeString.startsWith(".") || nodeString.startsWith("/"))
		) {
			source = nodeString;
		} else {
			source = tg.Graph.Edge.fromDataString(nodeString, tg.Object.withId);
		}
		let options: tg.Referent.Options = {};
		if (params !== undefined) {
			for (let param of params.split("&")) {
				let [key, value] = param.split("=");
				if (value === undefined) {
					throw new Error("missing value");
				}
				switch (key) {
					case "artifact": {
						options.artifact = decodeURIComponent(value);
						break;
					}
					case "id": {
						options.id = decodeURIComponent(value);
						break;
					}
					case "location": {
						options.location = tg.Location.fromDataString(
							decodeURIComponent(value),
						);
						break;
					}
					case "name": {
						options.name = decodeURIComponent(value);
						break;
					}
					case "path": {
						options.path = decodeURIComponent(value);
						break;
					}
					case "tag": {
						options.tag = decodeURIComponent(value);
						break;
					}
					case "kind": {
						kind = decodeURIComponent(value) as Kind;
						break;
					}
					default: {
						let match = key?.match(/^tokens\[(.*)\]$/);
						if (match === null || match === undefined) {
							throw new Error("invalid key");
						}
						options.tokens ??= {};
						options.tokens[decodeURIComponent(match[1]!)] =
							decodeURIComponent(value);
					}
				}
			}
		}
		tg.assert(kind !== undefined);
		let module = new tg.Module({
			kind,
			referent: {
				node: source,
				options,
			},
		});
		return module;
	};

	export let children = (value: Module): Array<tg.Object> => {
		let children =
			typeof value.referent.node !== "string"
				? tg.Graph.Edge.children(value.referent.node)
				: [];
		for (let child of children) {
			tg.Object.inheritLocation(
				child,
				value.referent.options?.location ?? null,
			);
			tg.Object.inheritTokens(child, value.referent.options?.tokens ?? {});
		}
		return children;
	};

	export let withoutToken = (value: tg.Module): tg.Module => {
		return new tg.Module({
			kind: value.kind,
			referent: tg.Referent.withoutToken(value.referent),
		});
	};

	export type Data = {
		kind: Module.Kind;
		referent: tg.Referent.Data<tg.Graph.Data.Edge<tg.Object.Id>>;
	};

	export namespace Data {
		export let children = (data: tg.Module.Data): Array<tg.Object.Id> => {
			let source =
				typeof data.referent === "string" ? data.referent : data.referent.node;
			if (
				typeof source === "string" &&
				(source.startsWith(".") || source.startsWith("/"))
			) {
				return [];
			}
			return tg.Graph.Data.Edge.children(source);
		};

		export let withoutTokens = (data: tg.Module.Data): tg.Module.Data => {
			if (typeof data.referent === "string") {
				let referent = tg.Referent.fromDataString(
					data.referent,
					(source) => source,
				);
				return {
					...data,
					referent: tg.Referent.toDataString(
						tg.Referent.withoutRuntime(referent),
						(source) => source,
					),
				};
			}
			let referent = tg.Referent.fromData(data.referent, (source) => source);
			return {
				...data,
				referent: tg.Referent.toData(
					tg.Referent.withoutRuntime(referent),
					(source) => source,
				),
			};
		};
	}

	export type Location = {
		module: tg.Module;
		range: tg.Range;
	};

	export namespace Location {
		export type Data = {
			module: tg.Module.Data;
			range: tg.Range;
		};

		export let toData = (
			value: tg.Module.Location,
		): tg.Module.Location.Data => {
			return {
				module: tg.Module.toData(value.module),
				range: value.range,
			};
		};

		export let fromData = (
			data: tg.Module.Location.Data,
		): tg.Module.Location => {
			return {
				module: tg.Module.fromData(data.module),
				range: data.range,
			};
		};

		export let children = (value: tg.Module.Location): Array<tg.Object> => {
			return tg.Module.children(value.module);
		};

		export namespace Data {
			export let children = (
				data: tg.Module.Location.Data,
			): Array<tg.Object.Id> => {
				return tg.Module.Data.children(data.module);
			};

			export let withoutTokens = (
				data: tg.Module.Location.Data,
			): tg.Module.Location.Data => {
				return {
					...data,
					module: tg.Module.Data.withoutTokens(data.module),
				};
			};
		}
	}
}

File: packages/clients/rust/src/module.rs
use {crate::prelude::*, std::path::Path};

pub use self::{
	data::Module as Data,
	handle::{Location, Module as Handle, Source},
	import::Import,
};

pub mod data;
pub mod handle;
pub mod import;
pub mod load;
pub mod resolve;

/// The possible file names for the root module in a package.
pub const ROOT_MODULE_FILE_NAMES: &[&str] = &["tangram.js", "tangram.ts"];

/// The file name of a lockfile.
pub const LOCKFILE_FILE_NAME: &str = "tangram.lock";

#[derive(
	Clone,
	Copy,
	Debug,
	Eq,
	Hash,
	Ord,
	PartialEq,
	PartialOrd,
	derive_more::Display,
	derive_more::FromStr,
	serde_with::DeserializeFromStr,
	serde_with::SerializeDisplay,
	tangram_serialize::Deserialize,
	tangram_serialize::Serialize,
)]
#[display(rename_all = "snake_case")]
#[from_str(rename_all = "snake_case")]
pub enum Kind {
	#[tangram_serialize(id = 0)]
	Js,

	#[tangram_serialize(id = 1)]
	Ts,

	#[tangram_serialize(id = 2)]
	Dts,

	#[tangram_serialize(id = 3)]
	Object,

	#[tangram_serialize(id = 4)]
	Artifact,

	#[tangram_serialize(id = 5)]
	Blob,

	#[tangram_serialize(id = 6)]
	Directory,

	#[tangram_serialize(id = 7)]
	File,

	#[tangram_serialize(id = 8)]
	Symlink,

	#[tangram_serialize(id = 9)]
	Graph,

	#[tangram_serialize(id = 10)]
	Command,

	#[tangram_serialize(id = 11)]
	Error,
}

pub fn module_kind_for_path(path: impl AsRef<Path>) -> tg::Result<tg::module::Kind> {
	let path = path.as_ref();
	let Some(name) = path.file_name().and_then(|name| name.to_str()) else {
		return Err(tg::error!(path = %path.display(), "invalid path"));
	};
	if name.ends_with(".d.ts") {
		Ok(tg::module::Kind::Dts)
	} else if name == "tangram.js" || name.ends_with(".tg.js") {
		Ok(tg::module::Kind::Js)
	} else if name == "tangram.ts" || name.ends_with(".tg.ts") {
		Ok(tg::module::Kind::Ts)
	} else {
		Err(tg::error!(path = %path.display(), "unknown or missing file extension"))
	}
}

#[must_use]
pub fn is_module_path(path: &Path) -> bool {
	let Some(name) = path.file_name() else {
		return false;
	};
	let Some(name) = name.to_str() else {
		return false;
	};
	tg::module::ROOT_MODULE_FILE_NAMES.contains(&name)
		|| name.ends_with(".tg.js")
		|| name.ends_with(".tg.ts")
}

#[must_use]
pub fn is_root_module_path(path: &Path) -> bool {
	let Some(name) = path.file_name() else {
		return false;
	};
	let Some(name) = name.to_str() else {
		return false;
	};
	tg::module::ROOT_MODULE_FILE_NAMES.contains(&name)
}

#[must_use]
pub fn is_non_root_module_path(path: &Path) -> bool {
	let Some(name) = path.file_name() else {
		return false;
	};
	let Some(name) = name.to_str() else {
		return false;
	};
	!tg::module::ROOT_MODULE_FILE_NAMES.contains(&name)
		&& (name.ends_with(".tg.js") || name.ends_with(".tg.ts"))
}

pub async fn try_get_root_module_file_name(
	package: tg::Either<&tg::Directory, &Path>,
) -> tg::Result<Option<&'static str>> {
	let handle = tg::handle()?;
	try_get_root_module_file_name_with_handle(handle, package).await
}

pub async fn try_get_root_module_file_name_with_handle<H>(
	handle: &H,
	package: tg::Either<&tg::Directory, &Path>,
) -> tg::Result<Option<&'static str>>
where
	H: tg::Handle,
{
	let mut name = None;
	for name_ in tg::module::ROOT_MODULE_FILE_NAMES {
		let exists = match package {
			tg::Either::Left(directory) => directory
				.try_get_entry_with_handle(handle, name_)
				.await
				.map_err(|error| tg::error!(!error, "failed to get the entry"))?
				.is_some(),
			tg::Either::Right(path) => tokio::fs::try_exists(path.join(*name_)).await.map_err(
				|error| tg::error!(!error, path = %path.display(), "failed to get the metadata"),
			)?,
		};
		if exists {
			if name.is_some() {
				return Err(tg::error!("package contains multiple root modules"));
			}
			name = Some(*name_);
		}
	}
	Ok(name)
}

pub fn try_get_root_module_file_name_sync(path: &Path) -> tg::Result<Option<&'static str>> {
	let mut name = None;
	for name_ in tg::module::ROOT_MODULE_FILE_NAMES {
		let exists = path.join(name_).try_exists().map_err(
			|error| tg::error!(!error, path = %path.display(), "failed to get the metadata"),
		)?;
		if exists {
			if name.is_some() {
				return Err(tg::error!("found multiple root modules"));
			}
			name = Some(*name_);
		}
	}
	Ok(name)
}
